src/devices/CharacterLcd/Hd44780.cs

```diff
@@ -0,0 +1,554 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Device.Gpio;
+using System.Diagnostics;
+using System.Drawing;
+using System.Threading;
+
+namespace Iot.Device.CharacterLcd
+{
+    /// <summary>
+    /// Supports LCD character displays compatible with the HD44780 LCD controller/driver.
+    /// Also supports serial interface adapters such as the MCP23008.
+    /// </summary>
+    /// <remarks>
+    /// The Hitatchi HD44780 was released in 1987 and set the standard for LCD controllers. Hitatchi does not make this chipset anymore, but
+    /// most character LCD drivers are intended to be fully compatible with this chipset. Some examples: Sunplus SPLC780D, Sitronix ST7066U,
+    /// Samsung KS0066U (and many more).
+    /// 
+    /// Some compatible chips extend the HD44780 with addtional pins and features. They are still fully compatible. The ST7036 is one example.
+    /// 
+    /// This implementation was drawn from numerous datasheets and libraries such as Adafruit_Python_CharLCD.
+    /// </remarks>
+    public class Hd44780 : IDisposable
+    {
+        private const byte ClearDisplayCommand = 0b_0001;
+        private const byte ReturnHomeCommand   = 0b_0010;
+
+        private const byte SetCGRamAddressCommand = 0b_0100_0000;
+        private const byte SetDDRamAddressCommand = 0b_1000_0000;
+
+        /// <summary>
+        /// Register select pin. Low is for writing to the instruction
+        /// register and reading the address counter. High is for reading
+        /// and writing to the data register.
+        /// </summary>
+        private readonly int _rsPin;
+
+        /// <summary>
+        /// Read/write pin. Low for write, high for read.
+        /// </summary>
+        private readonly int _rwPin;
+
+        /// <summary>
+        /// Enable pin. Pulse high to process a read/write.
+        /// </summary>
+        private readonly int _enablePin;
+
+        private readonly int _backlight;
+
+        private readonly int[] _dataPins;
+
+        private IGpioController _controller;
+
+        private DisplayFunction _displayFunction = DisplayFunction.Command;
+        private DisplayControl _displayControl = DisplayControl.Command;
+        private DisplayEntryMode _displayMode = DisplayEntryMode.Command;
+
+        private readonly byte[] _rowOffsets;
+        private Stopwatch _stopwatch = new Stopwatch();
+        private byte _lastByte;
+        private bool _useLastByte;
+
+        // We need to add PWM support to make this useful (to drive the VO pin).
+        // For now we'll just stash the value and use it to decide the initial
+        // backlight state.
+        private float _backlightBrightness;
+
+        /// <summary>
+        /// Logical size, in characters, of the LCD.
+        /// </summary>
+        public Size Size { get; }
```

Comment:
I'm not convinced we should use Size here. Every other framework takes in rows and cols, which is not necessarily the same as Width and Height. Personally I would just stick with the two ints representing the size instead.

Reply:
Console does width and height, which is where I got this from.

Reply:
I see. I still think that it might be worth for this particular one to use rows and cols as ints instead for people that is trying to port existing python code, It will just be easier for them to do it. I'll leave that up to you.

Reply:
They are ints in the Size (Width and Height). Terms are Width and Height for anything related that I could find. I can separate them out if you think that is better, but I don't think it is hard to make the correlation between columns and width. Unless there is a more compelling reason I think it is better to stay consistent with the .NET terminology (as with SetCursorPosition).